File: backend/internal/models/rh.go
package models

import "time"

type TermoResponsabilidade struct {
	ID             uint       `gorm:"primaryKey" json:"id"`
	SolicitacaoID  *uint      `gorm:"column:solicitacao_id" json:"solicitacao_id"`
	AssetID        uint       `gorm:"column:asset_id;not null" json:"asset_id"`
	UsuarioID      uint       `gorm:"column:usuario_id;not null" json:"usuario_id"`
	Status         string     `gorm:"type:varchar(30);default:'Pendente'" json:"status"`
	ConteudoTermo  string     `gorm:"type:text;not null" json:"conteudo_termo"`
	DataGeracao    time.Time  `gorm:"column:data_geracao;autoCreateTime" json:"data_geracao"`
	DataAssinatura *time.Time `gorm:"column:data_assinatura" json:"data_assinatura"`

	Solicitacao *Solicitacao `gorm:"foreignKey:SolicitacaoID" json:"solicitacao,omitempty"`
	Asset       *Asset       `gorm:"foreignKey:AssetID" json:"asset,omitempty"`
	Usuario     *User        `gorm:"foreignKey:UsuarioID" json:"usuario,omitempty"`
}

func (TermoResponsabilidade) TableName() string { return "termos_responsabilidade" }

// RHStatus is an HR-managed period in an employee's work calendar.  The
// effective status is calculated from the periods that are active right now,
// so a scheduled vacation never needs a manual status change on its first day.
type RHStatus struct {
	ID          uint       `gorm:"primaryKey" json:"id"`
	UsuarioID   uint       `gorm:"column:usuario_id;not null;index" json:"usuario_id"`
	Tipo        string     `gorm:"type:varchar(30);not null;index" json:"tipo"`
	Inicio      time.Time  `gorm:"column:inicio;not null;index" json:"inicio"`
	Fim         *time.Time `gorm:"column:fim;index" json:"fim"`
	Horas       *float64   `gorm:"column:horas" json:"horas"`
	Observacao  *string    `gorm:"type:text" json:"observacao"`
	CriadoPorID uint       `gorm:"column:criado_por_id;not null" json:"criado_por_id"`
	CreatedAt   time.Time  `gorm:"column:created_at;autoCreateTime" json:"created_at"`

	Usuario   *User `gorm:"foreignKey:UsuarioID" json:"usuario,omitempty"`
	CriadoPor *User `gorm:"foreignKey:CriadoPorID" json:"criado_por,omitempty"`
}

func (RHStatus) TableName() string { return "rh_status" }

// RHComunicado is an individual or company-wide message, independently kept
// from the general system notices so it can be shown only to its recipient.
type RHComunicado struct {
	ID          uint       `gorm:"primaryKey" json:"id"`
	UsuarioID   *uint      `gorm:"column:usuario_id;index" json:"usuario_id"`
	Titulo      string     `gorm:"not null" json:"titulo"`
	Mensagem    string     `gorm:"type:text;not null" json:"mensagem"`
	Inicio      time.Time  `gorm:"column:inicio;not null;index" json:"inicio"`
	Fim         *time.Time `gorm:"column:fim;index" json:"fim"`
	Ativo       bool       `gorm:"default:true" json:"ativo"`
	CriadoPorID uint       `gorm:"column:criado_por_id;not null" json:"criado_por_id"`
	CreatedAt   time.Time  `gorm:"column:created_at;autoCreateTime" json:"created_at"`

	Usuario   *User `gorm:"foreignKey:UsuarioID" json:"usuario,omitempty"`
	CriadoPor *User `gorm:"foreignKey:CriadoPorID" json:"criado_por,omitempty"`
}

func (RHComunicado) TableName() string { return "rh_comunicados" }

// RHComunicadoLeitura stores acknowledgement per recipient, including for
// company-wide messages, without mutating the message for other employees.
type RHComunicadoLeitura struct {
	ID           uint      `gorm:"primaryKey" json:"id"`
	ComunicadoID uint      `gorm:"column:comunicado_id;not null;uniqueIndex:idx_rh_comunicado_leitura" json:"comunicado_id"`
	UsuarioID    uint      `gorm:"column:usuario_id;not null;uniqueIndex:idx_rh_comunicado_leitura" json:"usuario_id"`
	LidoEm       time.Time `gorm:"column:lido_em;autoCreateTime" json:"lido_em"`
}

func (RHComunicadoLeitura) TableName() string { return "rh_comunicado_leituras" }

File: frontend/src/types/rh.ts
import type { Asset } from './asset';
import type { Solicitacao } from './transaction';

export interface TermoResponsabilidade {
  id: number;
  solicitacao_id?: number;
  solicitacao?: Solicitacao;
  asset_id: number;
  asset?: Asset;
  usuario_id: number;
  usuario?: {
    id: number;
    nome: string;
    email: string;
  };
  status: 'Pendente' | 'Assinado' | 'Cancelado' | string;
  conteudo_termo: string;
  data_geracao: string;
  data_assinatura?: string;
}

export interface RHListResponse {
  termos: TermoResponsabilidade[];
  pendentes: Solicitacao[];
}

export type RHStatusType = 'trabalhando' | 'folga' | 'ferias' | 'banco_horas' | 'desligado';

export interface RHStatusRecord {
  id: number;
  usuario_id: number;
  tipo: RHStatusType;
  inicio: string;
  fim?: string | null;
  horas?: number | null;
  observacao?: string | null;
  usuario?: { id: number; nome: string; email: string; cargo?: string | null; departamento_id?: number | null; departamento?: { id: number; nome: string } | null };
  criado_por?: { id: number; nome: string };
}

export interface RHComunicado {
  id: number;
  usuario_id?: number | null;
  titulo: string;
  mensagem: string;
  inicio: string;
  fim?: string | null;
  ativo: boolean;
  usuario?: { id: number; nome: string };
  criado_por?: { id: number; nome: string };
}

export interface RHControlResponse {
  colaboradores: Array<{ usuario: import('./user').User; status_atual: RHStatusType; horas?: number | null }>;
  status: RHStatusRecord[];
  comunicados: RHComunicado[];
  atualizado_em: string;
}

export interface RHMonitoringTeamResponse {
  colaboradores: Array<{ usuario: import('./user').User; status_atual: RHStatusType; horas?: number | null }>;
  atualizado_em: string;
}

export interface MyRHPortal {
  status_atual: RHStatusType;
  calendario: RHStatusRecord[];
  comunicados: Array<{ comunicado: RHComunicado; lida: boolean }>;
  atualizado_em: string;
}
